File: ios/RoomFinderAI/Models/RoommateProfile.swift
import Foundation

/// Someone on the roommate marketplace — either looking for a room, or with a
/// room to share.
struct RoommateProfile: Identifiable, Decodable, Hashable {

    enum Kind: String, Decodable {
        /// Looking for a room.
        case seeking
        /// Has a spot to fill.
        case hasSpot = "has_spot"

        var label: String {
            switch self {
            case .seeking: return "Looking for a room"
            case .hasSpot: return "Has a room"
            }
        }

        var symbol: String {
            switch self {
            case .seeking: return "magnifyingglass"
            case .hasSpot: return "house.fill"
            }
        }
    }

    let id: String
    let name: String?
    let userType: String?
    let budgetMin: Int?
    let budgetMax: Int?
    let preferredAreas: [String]?
    let moveInDate: String?
    let bio: String?
    let avatarUrl: String?
    let roomRent: Int?
    let roomLocation: String?
    let roomDescription: String?
    let roomPhotos: [String]?

    enum CodingKeys: String, CodingKey {
        case id, name, bio
        case userType = "user_type"
        case budgetMin = "budget_min"
        case budgetMax = "budget_max"
        case preferredAreas = "preferred_areas"
        case moveInDate = "move_in_date"
        case avatarUrl = "avatar_url"
        case roomRent = "room_rent"
        case roomLocation = "room_location"
        case roomDescription = "room_description"
        case roomPhotos = "room_photos"
    }

    var kind: Kind { Kind(rawValue: userType ?? "seeking") ?? .seeking }

    var displayName: String { name?.nilIfEmpty ?? "Someone" }

    /// Seed rows carry a `[seed]` marker that must never reach a real person.
    var cleanBio: String? {
        bio?.replacingOccurrences(of: "[seed]", with: "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .nilIfEmpty
    }

    /// What they're offering, or what they can pay. Only sensible values are
    /// shown: one seed row has a budget of $23–$333,333, which is noise.
    var budgetText: String {
        if kind == .hasSpot, let rent = roomRent, rent > 0 {
            return "$\(rent)/mo"
        }
        guard let low = budgetMin, let high = budgetMax,
              low > 0, high > low, high < 20_000 else {
            if let high = budgetMax, high > 0, high < 20_000 { return "Up to $\(high)/mo" }
            return "Budget not set"
        }
        return "$\(low)–$\(high)/mo"
    }

    var locationText: String {
        if let room = roomLocation?.nilIfEmpty { return room }
        if let areas = preferredAreas, !areas.isEmpty { return areas.joined(separator: ", ") }
        return "Anywhere"
    }

    var avatarURL: URL? {
        guard let avatarUrl, !avatarUrl.isEmpty else { return nil }
        return URL(string: avatarUrl)
    }

    /// Initials for the fallback avatar, so a profile without a photo still
    /// looks like a person rather than a broken image.
    var initials: String {
        let parts = displayName.split(separator: " ").prefix(2)
        let letters = parts.compactMap { $0.first }.map(String.init)
        return letters.isEmpty ? "?" : letters.joined().uppercased()
    }
}

struct RoommateProfilesResponse: Decodable {
    let success: Bool
    let data: [RoommateProfile]?
    let message: String?
}

File: RoomFinderAndroid/app/src/main/java/com/roomfinder/android/models/RoommateProfile.java
package com.roomfinder.android.models;

import com.google.gson.annotations.SerializedName;

import java.io.Serializable;
import java.util.List;
import java.util.Locale;

/**
 * Someone on the roommate marketplace: either looking for a room, or with a
 * room to share. Mirrors the iOS RoommateProfile, including the seed-row marker
 * that must never be shown to a real person.
 */
public class RoommateProfile implements Serializable {

    public static final String KIND_SEEKING = "seeking";
    public static final String KIND_HAS_SPOT = "has_spot";

    @SerializedName("id")
    private String id;

    @SerializedName("name")
    private String name;

    @SerializedName("user_type")
    private String userType;

    @SerializedName("budget_min")
    private Integer budgetMin;

    @SerializedName("budget_max")
    private Integer budgetMax;

    @SerializedName("preferred_areas")
    private List<String> preferredAreas;

    @SerializedName("move_in_date")
    private String moveInDate;

    @SerializedName("bio")
    private String bio;

    @SerializedName("avatar_url")
    private String avatarUrl;

    @SerializedName("room_rent")
    private Integer roomRent;

    @SerializedName("room_location")
    private String roomLocation;

    @SerializedName("room_description")
    private String roomDescription;

    @SerializedName("room_photos")
    private List<String> roomPhotos;

    public String getId() {
        return id;
    }

    public String getKind() {
        return KIND_HAS_SPOT.equals(userType) ? KIND_HAS_SPOT : KIND_SEEKING;
    }

    public boolean hasSpot() {
        return KIND_HAS_SPOT.equals(getKind());
    }

    public String getKindLabel() {
        return hasSpot() ? "Has a room" : "Looking for a room";
    }

    public String getDisplayName() {
        return (name == null || name.trim().isEmpty()) ? "Someone" : name.trim();
    }

    /** Seed rows carry a marker; they are not real people and are filtered out. */
    public boolean isSeed() {
        String haystack = ((name == null ? "" : name) + " " + (bio == null ? "" : bio)).toLowerCase();
        return haystack.contains("[seed]");
    }

    public String getInitials() {
        String display = getDisplayName();
        StringBuilder initials = new StringBuilder();
        for (String part : display.split("\\s+")) {
            if (!part.isEmpty() && initials.length() < 2) {
                initials.append(Character.toUpperCase(part.charAt(0)));
            }
        }
        return initials.length() == 0 ? "?" : initials.toString();
    }

    public String getLocationText() {
        if (hasSpot() && roomLocation != null && !roomLocation.trim().isEmpty()) {
            return roomLocation.trim();
        }
        if (preferredAreas != null) {
            for (String area : preferredAreas) {
                if (area != null && !area.trim().isEmpty()) {
                    return capitalise(area.trim());
                }
            }
        }
        return "Anywhere";
    }

    /** "$700 - $1,000/mo", "$900/mo", or "Budget not set". */
    public String getBudgetText() {
        if (hasSpot() && roomRent != null && roomRent > 0) {
            return String.format(Locale.US, "$%,d/mo", roomRent);
        }
        if (budgetMin != null && budgetMax != null && budgetMax > 0) {
            return String.format(Locale.US, "$%,d - $%,d/mo", budgetMin, budgetMax);
        }
        if (budgetMax != null && budgetMax > 0) {
            return String.format(Locale.US, "Up to $%,d/mo", budgetMax);
        }
        if (budgetMin != null && budgetMin > 0) {
            return String.format(Locale.US, "From $%,d/mo", budgetMin);
        }
        return "Budget not set";
    }

    public String getBio() {
        if (hasSpot() && roomDescription != null && !roomDescription.trim().isEmpty()) {
            return roomDescription.trim();
        }
        return bio == null ? "" : bio.trim();
    }

    public String getAvatarUrl() {
        return avatarUrl;
    }

    public String getMoveInDate() {
        return moveInDate;
    }

    public List<String> getRoomPhotos() {
        return roomPhotos;
    }

    private static String capitalise(String value) {
        if (value.isEmpty()) {
            return value;
        }
        return Character.toUpperCase(value.charAt(0)) + value.substring(1);
    }
}
